Update `github.com/moby/sys/user` and use of idtools

dmcgowan opened the pull request:
Update to use idtools equivalent functionality from `github.com/moby/sys/user`.

Not all functions are moved over so I expect deprecation failures

This shouldn't be a prerequisite for moving archive out. Since archive exposes idtools in its interface, the archive will be easier to do when the package name changes.

dmcgowan committed:
Update moby/sys/user to version which includes mapping

Update idtools to use Mkdir funcs from moby sys/user package
Add deprecation exception to golanci until move off idtools is complete

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Update daemon to use moby sys/user identity mapping

Signed-off-by: Derek McGowan <derek@mcg.dev>

thaJeztah commented:
Looks like there's a compile failure;

```
=== Errors
internal/usergroup/add_linux_test.go:33:36: undefined: mobyuser.GetRootUIDGID
```

thaJeztah commented on pkg/idtools/idtools.go:
```diff
@@ -3,11 +3,14 @@ package idtools
 // IDMap contains a single entry for user namespace range remapping. An array
 // of IDMap entries represents the structure that will be provided to the Linux
 // kernel for creating a user namespace.
+// Deprecated: use user.IDMap instead
```
These need an empty line before them, otherwise they're not picked up as a deprecation comment. When updating, it's also good to make the replacement a docs-link (adding square brackets around them); those are rendered as links to the replacement on pkg.go.dev.

```suggestion

// Deprecated: use [user.IDMap] instead
```

thaJeztah commented on pkg/idtools/idtools.go:
```diff
@@ -101,6 +108,26 @@ type IdentityMapping struct {
 	GIDMaps []IDMap `json:"GIDMaps"`
 }
 
+// FromUserIdentityMapping converts a user.IdentityMapping to an idtools.IdentityMapping
+func FromUserIdentityMapping(u user.IdentityMapping) IdentityMapping {
```
This utility is probably only needed for existing code to migrate, correct? Or do we want to continue having the `IdentityMapping` type?

If not, we could consider marking both this function and the type as deprecated, outlining that it's kept to help codebases to transition, but should not be used for new code. Something like;


```suggestion
// FromUserIdentityMapping converts a [user.IdentityMapping] to an [idtools.IdentityMapping].
//
// Deprecated: this is a transitional utility, and should not be used for reasons other than migrating existing code to [user.IdentityMapping].
```

And similar for the type (GitHub doesn't let me comment on those lines);


```go
// IdentityMapping contains a mappings of UIDs and GIDs.
// The zero value represents an empty mapping.
//
// Deprecated: use [user.IdentityMapping]  instead. This type is provided to allow existing code to transition.
type IdentityMapping struct {
	UIDMaps []IDMap `json:"UIDMaps"`
	GIDMaps []IDMap `json:"GIDMaps"`
}
```

dmcgowan replied:
`idtools.IdentityMapping` is still heavily used, trying to avoid deprecating it in this PR. I was thinking it would be better todo that after splitting archive

thaJeztah commented on vendor.mod:
```diff
@@ -256,3 +256,5 @@ exclude (
 	github.com/pmezard/go-difflib v1.0.1-0.20181226105442-5d4384ee4fb2
 )
+
+replace github.com/moby/sys/user v0.3.0 => github.com/moby/sys/user v0.0.0-20250227193707-71f0c5ead442
```
@dmcgowan I tagged v0.4.0, so you can remove this replace rule

I was about to open a PR to update the version, but we probably want to do that as part of this PR so that it's clear that files were moved from here to the other module.

dmcgowan commented:
The rest of the switch will come with the archive switch. That will be in a subsequent PR

thaJeztah committed:
pkg/idtools: remove tests already covered in moby/sys/user

Removes all tests, except for TestGetRootUIDGID and TestToContainer, which
are the only once that have a local implementation that's not covered.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on vendor/github.com/moby/sys/user/idtools_unix.go:
The "add" and "remove" are currently in 2 separate commits; I think we should try to make that an atomic change and have them in the same commit, so that it more properly shows up as a lateral move.

Perhaps organise the commits;


#### First commit (squash of 1st, 2nd, and 4th commit);

- "Update moby/sys/user to version which includes mapping"
- "Update idtools to use Mkdir funcs from moby sys/user package"
- "Add deprecation exception to golanci until move off idtools is complete"

#### Second commit (mosty to not make the first commit overly large I guess):

- "Update daemon to use moby sys/user identity mapping"

thaJeztah commented on pkg/idtools/idtools.go:
```diff
@@ -17,27 +21,35 @@ type IDMap struct {
+	return user.MkdirAllAndChown(path, mode, owner.UID, owner.GID, user.WithOnlyNew)
 }
 
 // GetRootUIDGID retrieves the remapped root uid/gid pair from the set of maps.
 // If the maps are empty, then the root uid/gid will default to "real" 0/0
+//
+// Deprecated: use [(user.IdentityMapping).RootPair] instead
 func GetRootUIDGID(uidMap, gidMap []IDMap) (int, int, error) {
```
This one is still used internal; could we split it into an internal (implementation) and the exported one (deprecated)? It makes it clearer that it's no longer used in our own code;


```go
// GetRootUIDGID retrieves the remapped root uid/gid pair from the set of maps.
// If the maps are empty, then the root uid/gid will default to "real" 0/0
//
// Deprecated: use [(user.IdentityMapping).RootPair] instead.
func GetRootUIDGID(uidMap, gidMap []IDMap) (int, int, error) {
	return getRootUIDGID(uidMap, gidMap)
}

// getRootUIDGID retrieves the remapped root uid/gid pair from the set of maps.
// If the maps are empty, then the root uid/gid will default to "real" 0/0
func getRootUIDGID(uidMap, gidMap []IDMap) (int, int, error) {
	uid, err := toHost(0, uidMap)
...
```


Then update the internal uses to use the non-exported one.

thaJeztah reviewed (changes requested):
Thanks! Overall looks good, but I left some suggestions; I also;

- opened a PR against this branch with my suggestions; https://github.com/dmcgowan/docker/pull/60
- pushed a quick branch with my "squash" considerations; https://github.com/moby/moby/pull/49744

thaJeztah reviewed (approved):
LGTM, thanks!
